inc/azure_macro_utils/macro_utils.h

```diff
@@ -78,73 +78,100 @@ MU_IF(X, "true", "false") => "true"
 #define MU_EAT_EMPTY_ARG(arg_count, arg) MU_IF(MU_ISEMPTY(arg),,arg) MU_IF(MU_ISEMPTY(arg),MU_EXPAND_TO_NOTHING,MU_IFCOMMALOGIC)(MU_DEC(arg_count))
 #define MU_EAT_EMPTY_ARGS(...) MU_FOR_EACH_1_COUNTED(MU_EAT_EMPTY_ARG, __VA_ARGS__)
 
+#define MU_EMPTY()
+#define MACRO_UTILS_DELAY(id) id MU_EMPTY MU_LPAREN )
+
+//
+//
+// MU_DEFINE_ENUM
+//
+//
+
+// This uber enum macro can handle enumeration-constants being constant expressions.
+// The ToString expands to something like the below
+
+// typedef enum TEST_UBER_ENUM_TAG
+// {
+//     test_uber_a,
+//     test_uber_b = 5
+// } TEST_UBER_ENUM;
+// 
+// MU_DEFINE_UBER_ENUM(TEST_UBER_ENUM, TEST_UBER_ENUM_VALUES)
+// 
+// const char* x_ToString(TEST_UBER_ENUM value)
+// {
+//     static int my_test_uber_a;
+//     static int my_test_uber_b = 5;
+//     static int enum_last_value = -1;
+//     enum_last_value = my_test_uber_a ? (my_test_uber_a) : (enum_last_value + 1);
```

Comment:
>enum_last_value = my_test_uber_a ? (my_test_uber_a) : (enum_last_value + 1); [](start = 7, length = 76)

I just thought of one edge case:

```
#define FOO_VALUES \
  foo_a = 1, \
  foo_b = 0

MU_DEFINE_ENUM(FOO, FOO_VALUES);

MU_ENUM_TO_STRING(FOO, foo_b) //= "NULL"?
MU_ENUM_TO_STRING(FOO, 0) //= "NULL"?
```

Reply:
Added some tests

Reply:
Tests added for it and now it works :-).

---
In reply to: [328298401](https://github.com/Azure/azure-macro-utils-c/pull/30#discussion_r328298401) [](ancestors = 328298401)